src/expr/attribute_internals.h

```diff
@@ -145,15 +149,383 @@ inline uint64_t GetBitSet(uint64_t bit)
+
+    /** Iterator within the second level IdMap (sorted vector) */
+    L2It d_l2It;
+  };
+
+  using Storage = std::unordered_map<NodeValue*, IdMap, AttrBoolHashFunction>;
+
+ public:
+  using iterator = Iterator<AttrHash<V>,
+                            typename Storage::iterator,
+                            typename IdMap::iterator>;
+  using const_iterator = Iterator<const AttrHash<V>,
+                                  typename Storage::const_iterator,
+                                  typename IdMap::const_iterator>;
+
+  std::size_t size() const
+  {
+    return std::accumulate(
+        d_storage.begin(),
+        d_storage.end(),
+        0u,
+        [](std::size_t sum, const std::pair<NodeValue*, IdMap>& l2) {
+          return sum + l2.second.size();
+        });
+  }
+
+  iterator begin() { return iterator(this); }
+  iterator end() { return iterator(); }
+
+  const_iterator begin() const
+  {
+    return const_iterator(const_cast<AttrHash<V>*>(this));
+  }
+  const_iterator end() const { return const_iterator(); }
+
+  iterator erase(iterator it)
+  {
+    // reach inside the iterator to get L1/L2 positions
+    auto nextL2It = it.d_l1It->second.erase(it.d_l2It);
+
+    iterator nextIt(this, it.d_l1It, nextL2It);
+    nextIt.legalize();
+
+    if (it.d_l1It->second.empty())
+    {
+      // this erase has made the L2 map empty. delete it:
+      d_storage.erase(it.d_l1It);
+    }
+
+    return nextIt;
+  }
+
+  template <typename Iter>
+  void insert(Iter beg, Iter end)
+  {
+    using Entry = typename std::iterator_traits<Iter>::value_type;
+    using Entries = std::vector<Entry>;
+    using EntryIt = typename Entries::iterator;
+    Entries entries(beg, end);
+
+    // sort by second (NodeValue*) then first (uint64_t)
+    std::sort(
+        entries.begin(), entries.end(), [](const Entry& a, const Entry& b) {
+          return (a.first.second < b.first.second)
+                 || ((a.first.second == b.first.second)
+                     && (a.first.first < b.first.first));
+        });
+
+    auto find_different_nv = [](NodeValue* nv, EntryIt first, EntryIt last) {
+      return std::find_if(
+          first, last, [nv](const Entry& a) { return a.first.second != nv; });
+    };
+
+    std::size_t l1_unique_count = 0;
+    auto last = entries.end();
+    for (EntryIt it = entries.begin(); it != last;
+         it = find_different_nv(it->first.second, it, entries.end()))
+    {
+      ++l1_unique_count;
+    }
+    d_storage.reserve(d_storage.size() + l1_unique_count);
+
+    for (EntryIt it = entries.begin(); it != last;)
+    {
+      EntryIt chunk_end = std::find_if(it, entries.end(), [it](const Entry& v) {
+        return v.first.second != it->first.second;
+      });
+      // add to l2 map
+      auto& l2 = d_storage[it->first.second];
+      l2.reserve(l2.size() + std::distance(it, chunk_end));
+      for (; it != chunk_end; ++it)
+      {
+        l2.emplace(it->first.first, it->second);
+      }
+    }
+  }
```

Comment:
It'd be good if you add a few comments about what this block of code does. It becomes kind of clear when you read the code, but I think it may help others to understand the code more quickly.

Reply:
In retrospect... yes that's a bit sparse comment-wise!